feat: In band inventory updates

kasperkristensen opened the pull request:
**What**

- Adds an Inventory Service to confirm and adjust inventory quantity for product variants.
- Adds the use of the new Inventory Service to Order-, ProductVariant-, Cart-, Claim-, Return-, SwapService.
- Adds integrations test for all api endpoints that makes use of the above mentioned services.
- Adds unit tests for all the above mentioned services, that validate that inventory is always confirmed before an action and adjusted accordingly.

**Why**

- To prevent orders with items that are OOS.
- To better reflect changes in inventory due to orders, claims, swaps and returns.

**How**

- Adds a simple Inventory Service that can check and adjust the inventory of product variants.
- Checks inventory before orders, swaps and claims are created. Fails gracefully if variant is OOS or only has partial stock.
- Adjusts inventory_quantity when cart is completed, when a claims is created, a return is received, and a swap is created.

**Testing**

- integrations test for all api endpoints that makes use of the above mentioned services.
- unit tests for all the above mentioned services, that validate that inventory is always confirmed before an action and adjusted accordingly.

olivermrbl commented on packages/medusa/src/services/cart.js:
```diff
@@ -455,38 +439,24 @@ class CartService extends BaseService {
-        // Confirm inventory
-        const hasInventory = await this.confirmInventory_(
+        // Confirm inventory or throw error
+        await this.inventoryService_.confirmInventory(
```
No reason to check inventory in both branches of the `if`-statement. Let us move it outside just before. And check that `item.variant_id` exists too. It might be a custom item

olivermrbl replied:
And add `.withTransaction(manager)`

olivermrbl commented on packages/medusa/src/services/order.js:
```diff
@@ -550,6 +561,12 @@ class OrderService extends BaseService {
       }
 
+      for (const item of cart.items) {
+        await this.inventoryService_
```
check for `item.variant_id`

olivermrbl commented on packages/medusa/src/services/order.js:
```diff
@@ -881,6 +898,12 @@ class OrderService extends BaseService {
         )
       )
 
+      for (const item of order.items) {
```
check for `item.variant_id`

olivermrbl commented on packages/medusa/src/services/inventory.js:
```diff
@@ -0,0 +1,73 @@
+   */
+  async adjustInventory(variantId, adjustment) {
+    return this.atomicPhase_(async (manager) => {
+      const variant = await this.productVariantService_.retrieve(variantId)
```
I think we need to add a check, that ensures that the variant should actually be updated. If `manage_inventory` is set to false, we do not want to update it right?

olivermrbl commented on packages/medusa/src/services/return.js:
```diff
@@ -513,6 +517,15 @@ class ReturnService extends BaseService {
 
+      for (const line of newLines) {
+        const orderItem = order.items.find(i => i.id === line.item_id)
+        if (orderItem) {
```
check for `variant_id`

olivermrbl reviewed (changes requested):
Very good job!

A few changes needs to be made. Generally, we would like to add a check for `item.variant_id` before both confirming and adjusting. Otherwise, the retrieval of the variant would throw an error. 
I've added comments to most of the places, but since this would mean a lot of redundant code, we could investigate a more sophisticated solution to this.

Could be also add an integration tests, that ensures that no update will happen to a variant, if `variant.manage_inventory` is set to false.

kasperkristensen commented:
@olivermrbl Made the proposed changes 👍 

- Check if variant_id is undefined before retrieve
- Added check to adjustInventory so it skips variants that are not managed
- Added unit tests + integration tests to make sure that quantity is never adjusted for unmanaged variants

olivermrbl commented on package.json:
```diff
@@ -36,7 +36,9 @@
   "dependencies": {
+    "global": "^4.4.0",
     "import-from": "^3.0.0",
+    "medusa-dev-cli": "^0.0.15",
```
Are these intentionally added? Otherwise delete.

olivermrbl commented on packages/medusa/src/services/cart.js:
```diff
@@ -450,43 +434,22 @@ class CartService extends BaseService {
       }
 
+      // Confirm inventory or throw error
+      await this.inventoryService_
```
Sorry, I take my previous comment back. Since the condition for current item would result in a new larger quantity, we need to have the check both places :) Please add.

olivermrbl commented on packages/medusa/src/services/cart.js:
```diff
@@ -541,7 +504,7 @@ class CartService extends BaseService {
 
       if (lineItemUpdate.quantity) {
-        const hasInventory = await this.confirmInventory_(
+        const hasInventory = await this.inventoryService_.confirmInventory(
```
Should be: `inventoryService_.withTransaction(manager).confirmInventory`

olivermrbl reviewed (changes requested):
Minor changes :)

srindom commented on packages/medusa/src/services/order.js:
```diff
@@ -450,6 +455,12 @@ class OrderService extends BaseService {
+      for (const item of cart.items) {
+        await this.inventoryService_
+          .withTransaction(manager)
+          .confirmInventory(item.variant_id, item.quantity)
```
We bred to do some sort of compensating action on the payment in case this fails. The payment should technically be authorized at this point so if the order creation fails we have payment that is not linked to an order. @olivermrbl What are your thoughts?

olivermrbl replied:
I agree! Let's make sure to `cancelPayment`, if this stage in cart completion fails @kasperkristensen

kasperkristensen replied:
@olivermrbl, @srindom  Made the requested changes and it now catches the error on no inventory and cancels the payment. Added a unit test to validate the outcome in `/__tests__/order.js` at `fails because an item does not have the required inventory`.

olivermrbl reviewed (dismissed):
Awesome! Let's get it in.. Please squash and merge the PR 🚀

srindom reviewed (approved):
Tested in a test environment - works like a charm 🎉  Great work!

srindom commented on integration-tests/api/__tests__/store/cart.js:
```diff
@@ -204,6 +213,49 @@ describe("/store/carts", () => {
+      const pay = await manager.findOne(Payment, {
+        where: { id: res.data.cart.payment.id },
+      });
+      console.error(pay);
```
We should probably do some sort of assertion here? :)

olivermrbl commented:
Additions: 
- Complete cart returns early if cart is already completed
- Code added to MedusaError to have a generic error message
